extensions/basic_auth/plugin.h

```diff
@@ -0,0 +1,93 @@
+/* Copyright 2020 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "extensions/common/context.h"
+#include "google/protobuf/util/json_util.h"
+
+#ifndef NULL_PLUGIN
+
+#include <assert.h>
+#define ASSERT(_X) assert(_X)
+
+#include "proxy_wasm_intrinsics.h"
+
+static const std::string EMPTY_STRING;
+
+#else
+
+#include "include/proxy-wasm/null_plugin.h"
+
+namespace proxy_wasm {
+namespace null_plugin {
+namespace BasicAuth {
+namespace Plugin {
+
+#endif
+
+// PluginRootContext is the root context for all streams processed by the
+// thread. It has the same lifetime as the worker thread and acts as target for
+// interactions that outlives individual stream, e.g. timer, async calls.
+class PluginRootContext : public RootContext {
+ public:
+  PluginRootContext(uint32_t id, std::string_view root_id)
+      : RootContext(id, root_id) {}
+  ~PluginRootContext() {}
+  bool onConfigure(size_t) override;
+
+  enum MATCH_TYPE { Prefix, Exact, Suffix };
+  struct BasicAuthConfigRule {
+    std::string request_path;
+    MATCH_TYPE pattern;
+    std::unordered_set<std::string> encoded_credentials;
+  };
```

Comment:
nit: These no longer need to be public right?

Reply:
okok they were public because `checkCredentials()` function is inside anonymous namespace. I'm going to place the struct and enum private and move the function outside the namespace and include it as a private method in PluginRootContext.

Reply:
ok, sgtm.